Fetch remote organization via action api

metaodi opened the pull request:
Organizations used to be returned by `/api/2/rest/group`, this is what the
old implementation used to fetch the information to create the remote
organization on the local instance of CKAN.

With this commit the Action API is used to fetch the same information.

Fixes #120

amercader commented on ckanext/harvest/harvesters/ckanharvester.py:
```diff
@@ -324,7 +337,14 @@ def import_stage(self,harvest_object):
-                                org = self._get_group(harvest_object.source.url, remote_org)
+                                try:
+                                    org = self._get_organization(harvest_object.source.url, remote_org)
+                                except:
```
@metaodi this is bad, see comment above

amercader commented on ckanext/harvest/harvesters/ckanharvester.py:
```diff
@@ -48,6 +52,15 @@ def _get_group(self, base_url, group_name):
+            content = self._get_content(url)
+            content_dict = json.loads(content)
+            return content_dict['result']
+        except Exception, e:
```
@metaodi this should be avoided, as it will catch Exceptions that are not related (eg if the json parsing failed). I guess that if we are targeting an old CKAN instance we will get a 404 if the action is not available, so we should only catch that (probably a urllib2 execption)

metaodi replied:
I understand the reasoning, but here it really doesn't matter why it fails. The implementation is analog to the implementation of `_get_group`, but of course it could be changed in both places. If - as you wrote - the parsing of the JSON failed, it should fail. Maybe it's better to remove the try/catch entirely, and delegate the exception handling to the caller.

In this particular case the [caller catches all exception in an `except` block](https://github.com/metaodi/ckanext-harvest/blob/120-create-remote-orgs/ckanext/harvest/harvesters/ckanharvester.py#L353), as any exception simply means, that the remote organization could not be fetched.

What do you think?

davidread replied:
There are also lots of programming errors that you can make, that cause different exceptions, which are nothing to do with the server's response. Not swallowing those exceptions helps fix the code. Hence why we have a CKAN policy of making the effort to identify the correct exception in except clauses. It's also mentioned in PEP8:

> When catching exceptions, mention specific exceptions whenever possible instead of using a bare except: clause.

amercader replied:
I hadn't noticed the line on 353. I still think that catch all exceptions are bad (even if wrote the `_get_group` one a long time ago :speak_no_evil: ) but I agree that we can let this one go in.

amercader replied:
But then again @davidread is right and we shouldn't do it, so let's fix `_get_group` and `_get_organization` to be more explicit :)

davidread commented:
PS Cheers for submitting this - most helpful

metaodi commented:
@davidread @amercader okay, I tried to clean this all up. I introduced two now *Error classes, one to wrap the urllib2 exception and one to wrap either the HTTP or the json parse error. I replaced all generic `except:` blocks and almost all generic `Exception` usages in the CKANHarvester.

During my tests I also found an [issue with the remote_groups](https://github.com/metaodi/ckanext-harvest/commit/935b9dda0100325028b01eb45a363b5730670c16), if you want me to create a separate PR for that, just let me know.

davidread commented on ckanext/harvest/harvesters/ckanharvester.py:
```diff
@@ -36,17 +41,34 @@ def _get_content(self, url):
+            raise ContentFetchError(
+                'Could not fetch url: %s, HTTP error code: %s' % 
+                (url, e.code)
+            )
```
I believe this also needs to catch urllib2.URLError, in cases such as where the url has bad syntax, or the socket connection times out.

metaodi replied:
Good catch, I'll replace it with urllib2.URLError (which is the super class of urllib2.HTTPError), this should cover the exceptions at this point.

davidread commented:
Thanks @metaodi - I think that's another really valuable improvement

davidread commented:
I didn't know about HTTPError inheriting URLError - cheers!

metaodi commented:
@davidread a thanks for this hint, replaced it with str() as you suggested.

davidread commented:
Thanks @metaodi
